metricbeat/module/kvm/dommemstat/dommemstat.go

```diff
@@ -0,0 +1,159 @@
+package dommemstat
+
+import (
+	"errors"
+	"net"
+	"net/url"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/metricbeat/mb"
+
+	"github.com/digitalocean/go-libvirt"
+	"github.com/digitalocean/go-libvirt/libvirttest"
+)
+
+const (
+	// maximum number of memory stats to be collected
+	// limit is defined by REMOTE_DOMAIN_MEMORY_STATS_MAX
+	// based on https://github.com/libvirt/libvirt/blob/5bb07527c11a6123e044a5dfc48bdeccee144994/src/remote/remote_protocol.x#L136
+	maximumStats = 11
+	// flag VIR_DOMAIN_AFFECT_CURRENT passed to collect memory stats
+	// based on https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainModificationImpact
+	flags = 0
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("kvm", "dommemstat", New)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	timeout time.Duration
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The kvm dommemstat metricset is experimental.")
+
+	config := struct{}{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		timeout:       base.Module().Config().Timeout,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) {
+
+	var (
+		c   net.Conn
+		err error
+	)
+
+	u, err := url.Parse(m.Host())
+	if err != nil {
+		report.Error(err)
```

Comment:
This parsing could be done in `New` so it is done only once. If it's kept here add a `return` after reporting the Error so it doesn't continue.